services/wallet/history/service.go

```diff
@@ -198,80 +201,178 @@ func (src *tokenChainClientSource) BalanceAt(ctx context.Context, account common
 // GetBalanceHistory returns token count balance
 // TODO: fetch token to FIAT exchange rates and return FIAT balance
 func (s *Service) GetBalanceHistory(ctx context.Context, chainIDs []uint64, address common.Address, currency string, endTimestamp int64, timeInterval TimeInterval) ([]*DataPoint, error) {
-	allData := make(map[uint64][]*DataPoint)
+	allData := make(map[chainIdentity][]*DataPoint)
 	for _, chainID := range chainIDs {
 		data, err := s.balance.get(ctx, chainID, currency, address, endTimestamp, timeInterval)
 		if err != nil {
 			return nil, err
 		}
 		if len(data) > 0 {
-			allData[chainID] = data
+			allData[chainIdentity(chainID)] = data
 		}
 	}
 
-	return mergeDataPoints(allData)
+	return mergeDataPoints(allData, strideDuration(timeInterval))
 }
 
-// mergeDataPoints merges same block numbers from different chains which are incompatible due to different timelines
-// TODO: use time-based intervals instead of block numbers
-func mergeDataPoints(data map[uint64][]*DataPoint) ([]*DataPoint, error) {
+// mergeDataPoints merges close in time block numbers. Drops the ones that are not in a stride duration
+// this should improve merging balance data from different chains which are incompatible due to different timelines
+// and block length
+func mergeDataPoints(data map[chainIdentity][]*DataPoint, stride time.Duration) ([]*DataPoint, error) {
 	if len(data) == 0 {
 		return make([]*DataPoint, 0), nil
 	}
 
-	pos := make(map[uint64]int)
+	pos := make(map[chainIdentity]int)
 	for k := range data {
 		pos[k] = 0
 	}
 
 	res := make([]*DataPoint, 0)
-	done := false
-	for !done {
-		var minNo *big.Int
-		var timestamp uint64
-		// Take the smallest block number
+	strideStart := findFirstStrideWindow(data, stride)
```

Comment:
The time interval for the entire data set is split into fixed time intervals based on the "stride" created by block numbers that we use in balance fetching.